blog/tidy-collaborative-filtering/index.qmd

```diff
@@ -0,0 +1,216 @@
+
+# Train Collaborative Filtering Recommenders in R
+
+For this post, we rely on the `tidyverse` [@Wickham2019] family of packages, `scales` [@scales] for scale functions for visualization, and `recommenderlab` [@recommenderlab] - a package that provides an infrastructure to develop and test collaborative filtering recommender algorithms.  
+
+```{r, message = FALSE, warning = FALSE} 
+library(tidyverse)
+library(scales)
+library(recommenderlab)
+```
+
+I load a data set with stock holdings of investable wikifolios at the beginning of 2023 that I host in one of my repositories. The data contains the portfolios of 6,544 wikifolios that held in total 5,069 stocks on January 1st 2023. 
+
+```{r, message = FALSE, warning = FALSE}
+wikifolio_portfolios <- read_csv("https://raw.githubusercontent.com/christophscheuch/christophscheuch.github.io/main/data/wikifolio_portfolios.csv") 
+glimpse(wikifolio_portfolios)
+```
+
+First, I convert the long data to a binary rating matrix. 
+
+```{r}
+binary_rating_matrix <- wikifolio_portfolios |>
+  mutate(in_portfolio = 1) |> 
+  pivot_wider(id_cols = wikifolio,
+              names_from = stock,
+              values_from = in_portfolio,
+              values_fill = list(in_portfolio = 0)) |>
+  select(-wikifolio) |>
+  as.matrix() |>
+  as("binaryRatingMatrix")
+binary_rating_matrix
+```
+
+As in our book chapter on [factor selection via machine learning](factor-selection-via-machine-learning.qmd), I perform cross-validation and split the data into training and test data. The training sample constitute 80% of the data and I perform 5-fold cross validation. Testing is performed by withholding items from the test portfolios (parameter `given`) and checking how well the algorithm predicts the withheld items. The `given=-1` means that an algorithm sees all but 1 withheld stock for the test portfolios and needs to predict the missing stock. I refer to @BreeseEtAl1998 for a discussion of other withholding strategies. 
+
+```{r}
+scheme <- binary_rating_matrix |>
+  evaluationScheme(
+    method = "cross-validation",
+    k      = 5,
+    train  = 0.8,
+    given  = -1
+)
+scheme
+```
+
+Here is the list of recommenders that I consider for the backtest with some intuition:
+
+* Random Items: the benchmark case because it just stupidly chooses random stocks from all possible choices.
+* Popular Items: just recommended the most popular stocks to measured by the number of wikifolios that hold the stock.
+* Association Rules: each wikifolio and its portfolio is considered as a transaction. Association rule mining finds similar portfolios across all traders (if a traders has x, y and z in his/her portfolio, then he/she is X% likely of also including w). 
```

Comment:
Should we use "they/them" as a neutral choice for including non-binary?

Also some typos "(if a trader has x, y, and z in their portfolio, then they are X% likely also to include w)"

Reply:
They / them works better in this case